### PR TITLE
optimize BFloat16 elemwise operators CPU: sigmoid, sigmoid_backward, tanh_backward, addcmul, addcdiv

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61015 add BFloat16 support for BatchNorm2d on CPU
* #58297 add BFloat16 support for UpSample on CPU
* #56904 optimize transpose copy for float32 and bfloat16 on CPU
* #56903 add BFloat16 support for MaxPool2d on CPU
* #56902 add BFloat16 support for AdaptiveAvgPool2d on CPU
* #56372 add BFloat16 support for bernoulli and Dropout on CPU
* #55588 add bf16 support for bucketize
* **#55221 optimize BFloat16 elemwise operators CPU: sigmoid, sigmoid_backward, tanh_backward, addcmul, addcdiv**

Differential Revision: [D28836797](https://our.internmc.facebook.com/intern/diff/D28836797)

### CODE REVIEW

**Comment by mingfeima** [comment]
Since this PR is not related to parallelization feature, only single core perf is tested:
NB: tanh_backward doesn't have BFloat16 support previously, the `before` perf refers to simple impl as following:
```diff
-    AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "tanh_backward_cpu", [&]() {
+    AT_DISPATCH_FLOATING_TYPES_AND(kBFloat16, iter.dtype(), "tanh_backward_cpu", [&]() {
```

- performance update on **avx512** machine: Xeon(R) Gold 6248 CPU @ 2.50GHz
```bash
before: sigmoid: 1x128x1024: fp32: 0.101 ms; bf16: 0.272 ms
after:  sigmoid: 1x128x1024: fp32: 0.109 ms; bf16: 0.144 ms

before: sigmoid backward: 1x128x1024: fp32: 0.114 ms; bf16: 0.154 ms
after:  sigmoid backward: 1x128x1024: fp32: 0.111 ms; bf16: 0.091 ms

before: tanh backward: 1x128x1024: fp32: 0.110 ms; bf16: 0.185 ms
after:  tanh backward: 1x128x1024: fp32: 0.111 ms; bf16: 0.085 ms
```
- performance update on **avx2** machine: Xeon(R) CPU E5-2680 v3 @ 2.50GHz
```bash
before: sigmoid: 1x128x1024: fp32: 0.241 ms; bf16: 0.480 ms
after:  sigmoid: 1x128x1024: fp32: 0.237 ms; bf16: 0.238 ms

before: sigmoid backward: 1x128x1024: fp32: 0.128 ms; bf16: 0.299 ms
after:  sigmoid backward: 1x128x1024: fp32: 0.131 ms; bf16: 0.149 ms

before: tanh backward: 1x128x1024: fp32: 0.134 ms; bf16: 0.299 ms
after:  tanh backward: 1x128x1024: fp32: 0.136 ms; bf16: 0.131 ms
```

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836797).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836797).

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 1]
```diff
@@ -673,6 +673,24 @@ void huber_kernel(TensorIterator& iter, double delta) {
 
 void sigmoid_backward_kernel(TensorIterator& iter) {
   AT_DISPATCH_FLOATING_TYPES_AND2(kBFloat16, kHalf, iter.dtype(), "sigmoid_backward_cpu", [&]() {
+    if (iter.dtype() == kBFloat16) {
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:676]: This looks weird, can you please do `if` statement otside of dispatch

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 2]
```diff
@@ -748,7 +766,27 @@ void tanh_backward_kernel(TensorIterator& iter) {
       });
   });
   } else {
-    AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "tanh_backward_cpu", [&]() {
+    AT_DISPATCH_FLOATING_TYPES_AND(kBFloat16, iter.dtype(), "tanh_backward_cpu", [&]() {
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:769]: Same as above

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Please change code pattern from dispatching on all types and specializing on bfloat16 to check type first and dispatch only on right types.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 3]
```diff
@@ -673,6 +673,24 @@ void huber_kernel(TensorIterator& iter, double delta) {
...
+    if (iter.dtype() == kBFloat16) {
+      auto one_vec = Vectorized<float>((float)(1));
+      cpu_kernel_vec(iter,
+        [=](BFloat16 a, BFloat16 b) -> BFloat16 {
+          float a0 = static_cast<float>(a);
+          float b0 = static_cast<float>(b);
+          return a0 * (float(1) - b0) * b0;
+        },
+        [=](Vectorized<BFloat16> a, Vectorized<BFloat16> b) {
+          Vectorized<float> a0, a1, b0, b1;
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:685]: This code looks like good candidate to have inside of vec library

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836797).

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 1 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: Yeah... good idea. Fixed!

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 2 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: Fixed!

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 3 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: Fixed!

**Comment by mingfeima** [comment]
rebased!

**Comment by mingfeima** [comment]
rebased!

**Comment by mingfeima** [comment]
rebased!

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by mingfeima** [comment]
rebased!

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836797).

**Pull request closed without merging**